internal/fs/buffer/inmemory_write_buffer.go

```diff
@@ -15,33 +15,117 @@
 package buffer
 
 import (
-	"bytes"
+	"fmt"
 )
 
+// InMemoryWriteBuffer implements WriteBuffer and is stored in memory.
 type InMemoryWriteBuffer struct {
-	// Holds the data to be written to GCS. At any time, Buffer holds the data for
-	// 2 GCS write calls.
-	buffer *bytes.Buffer
+	// Holds the incoming data from kernel write calls.
+	currentBuffer []byte
+	// [ currentBufferStartOffset, currentBufferEndOffset) is the range of file offset for which the
+	// data is currently stored in currentBuffer.
+	currentBufferStartOffset int64
+	currentBufferEndOffset   int64
+
+	// Holds the data that is sent to go storage client for upload. Data in this buffer
+	// will persist until the entire data is successfully uploaded to GCS.
+	flushedBuffer []byte
+	// [ flushedBufferStartOffset, flushedBufferEndOffset) is the range of file offset for which the
+	// data is currently stored in flushedBuffer.
+	flushedBufferStartOffset int64
+	flushedBufferEndOffset   int64
+
+	// Holds the last offset to which data is written.
+	fileSize int64
 }
 
 // CreateInMemoryWriteBuffer creates a buffer with InMemoryWriteBuffer.buffer
 // set to nil. Memory is allocated to the buffer when the first write call comes.
 // This avoids unnecessarily bloating GCSFuse memory consumption.
 //
-// To allocate memory to the buffer, use WriteBuffer.InitializeBuffer.
+// To allocate memory to the buffer, use WriteBuffer.Initialize.
 func CreateInMemoryWriteBuffer() *InMemoryWriteBuffer {
 	b := &InMemoryWriteBuffer{}
 	// TODO: set mtime attribute.
 	return b
 }
 
-func (b *InMemoryWriteBuffer) InitializeBuffer(sizeInMB int) {
-	ChunkSize = sizeInMB * MiB
-	if b.buffer == nil {
-		b.buffer = bytes.NewBuffer(make([]byte, 0, 2*sizeInMB*MiB))
+func (b *InMemoryWriteBuffer) Initialize(sizeInMB int) {
+	if ChunkSize == 0 {
+		ChunkSize = int64(sizeInMB * MiB)
+	}
+	if b.currentBuffer == nil {
+		b.currentBuffer = make([]byte, 0, ChunkSize)
+		b.currentBufferEndOffset = ChunkSize
+	}
+	if b.flushedBuffer == nil {
+		b.flushedBuffer = make([]byte, 0, ChunkSize)
+	}
+}
+
+func (b *InMemoryWriteBuffer) WriteAt(content []byte, receivedContentStartOffset int64) error {
```

Comment:
I think `offset` would do for the argument name as the rest is clear from the context.

```suggestion
func (b *InMemoryWriteBuffer) WriteAt(content []byte, offset int64) error {
```